File: client/EncryptedFileTransferClinet/CryptoManager.cpp
#include <iostream>
#include <string>
#include <cryptopp/rsa.h>
#include <cryptopp/osrng.h>
#include <cryptopp/aes.h>
#include <cryptopp/modes.h>
#include <cryptopp/filters.h>
#include <cryptopp/hex.h>
#include <cryptopp/crc.h>
#include <cryptopp/base64.h>
#include "CryptoManager.h"
#include <cryptopp/files.h>
#include <cryptopp/cryptlib.h>

using namespace CryptoPP;

// Constructor
CryptoManager::CryptoManager() : aesKey(AES_KEY_SIZE) {} // Changed to 32 bytes

void CryptoManager::generateRSAKeys()
{
        AutoSeededRandomPool rng;
        InvertibleRSAFunction params;
        params.GenerateRandomWithKeySize(rng, 1024);

        privateKey = RSA::PrivateKey(params);
        publicKey = RSA::PublicKey(params);
}

    //std::string CryptoManager::getEncryptedAESKey()
    //{
    //    return encryptRSA(aesKey);
    //}

    //TODO need to remove this function
    void CryptoManager::generateAESKey()
    {
        AutoSeededRandomPool rng;
        aesKey.resize(AES_KEY_SIZE); // Ensure the key is 32 bytes
        rng.GenerateBlock(aesKey, aesKey.size());
    }

    std::vector<uint8_t> CryptoManager::encryptAES(const std::vector<uint8_t>& data) {
        // Generate IV
        AutoSeededRandomPool prng;
        byte iv[AES::BLOCKSIZE];
        prng.GenerateBlock(iv, AES::BLOCKSIZE);

        // Set up encryption objects
        CBC_Mode<AES>::Encryption encryptor;
        encryptor.SetKeyWithIV(aesKey, AES_KEY_SIZE, iv);

        // Encrypt
        std::string ciphertext;
        StringSource(
            reinterpret_cast<const byte*>(data.data()), data.size(), true,
            new StreamTransformationFilter(encryptor,
                new StringSink(ciphertext)
            )
        );

        // Prepare result: IV + Ciphertext
        std::vector<uint8_t> result(iv, iv + AES::BLOCKSIZE);
        result.insert(result.end(), ciphertext.begin(), ciphertext.end());

        return result;
    }

    std::string CryptoManager::decryptAES(const std::vector<uint8_t>& data) {
        if (data.size() < AES::BLOCKSIZE) {
            throw std::invalid_argument("Input data is too short");
        }

        // Extract IV
        byte iv[AES::BLOCKSIZE];
        std::copy(data.begin(), data.begin() + AES::BLOCKSIZE, iv);

        // Set up decryption objects
        CBC_Mode<AES>::Decryption decryptor;
        decryptor.SetKeyWithIV(aesKey, AES_KEY_SIZE, iv);

        // Decrypt
        std::string plaintext;
        StringSource(
            reinterpret_cast<const byte*>(data.data()) + AES::BLOCKSIZE, 
            data.size() - AES::BLOCKSIZE, 
            true,
            new StreamTransformationFilter(decryptor,
                new StringSink(plaintext)
            )
        );

        // Convert result to vector<uint8_t>
        return plaintext;
    }

    std::string CryptoManager::encryptRSA(const std::string plainText)
    {
        SecByteBlock data(reinterpret_cast<const byte*>(plainText.data()), plainText.size());
        AutoSeededRandomPool rng;
        std::string ciphertext;

        RSAES_OAEP_SHA_Encryptor encryptor(publicKey);
        StringSource ss(data.data(), data.size(), true,
            new PK_EncryptorFilter(rng, encryptor,
                new StringSink(ciphertext)
            )
        );

        return ciphertext;
    }

    std::string CryptoManager::decryptRSA(const std::vector<uint8_t>& encryptedData) {
        AutoSeededRandomPool rng;
        std::string decryptedData;

        RSAES_OAEP_SHA_Decryptor decryptor(privateKey);
        StringSource ss(reinterpret_cast<const byte*>(encryptedData.data()), encryptedData.size(), true,
            new PK_DecryptorFilter(rng, decryptor,
                new StringSink(decryptedData)
            )
        );

        return decryptedData;
    }

    /*std::string CryptoManager::getCRC(const std::string& data, bool decrypt = false) {
        std::string processedData = data;
        if (decrypt) {
            processedData = decryptAES(data);
        }

        CRC32 crc;
        std::string hash;
        StringSource ss(processedData, true,
            new HashFilter(crc,
                new HexEncoder(
                    new StringSink(hash)
                )
            )
        );

        return hash;
    }*/
    // get the keys
    RSA::PrivateKey CryptoManager::getPrivateKey() {
		return privateKey;
	}

    std::vector<uint8_t> CryptoManager::getPublicKey() {
        ByteQueue queue;
        publicKey.DEREncodePublicKey(queue);

        std::vector<uint8_t> publicKeyVec(queue.CurrentSize());
        queue.Get(reinterpret_cast<byte*>(&publicKeyVec[0]), publicKeyVec.size());

        return publicKeyVec;
    }

    SecByteBlock CryptoManager::getAESKey() {
		return aesKey;
	}

    void CryptoManager::setAESKey(const std::string& keyString) 
    {
        SecByteBlock data(reinterpret_cast<const byte*>(keyString.data()), keyString.size());
        if (keyString.length() != AES_KEY_SIZE) {
            throw std::invalid_argument("Invalid AES key length");
        }

        aesKey = data;
    }

    /*void setAESKey(const SecByteBlock& key) {
        CryptoManager::aesKey = key;
	}*/



//int main() {
//    CryptoManager cryptoManager;
//    cryptoManager.generateRSAKeys();
//    cryptoManager.generateAESKey();
//    std::string encAESKey = cryptoManager.getEncryptedAESKey();
//    // get the keys as RSA::PrivateKey and RSA::PublicKey format and convert them to string and print them
//    RSA::PrivateKey privateKey = cryptoManager.getPrivateKey();
//    RSA::PublicKey publicKey = cryptoManager.getPublicKey();
//    std::string strPrivateKey, strPublicKey;
//    privateKey.Save(StringSink(strPrivateKey).Ref());
//    publicKey.Save(StringSink(strPublicKey).Ref());
//    std::cout << "Private Key:\n " << strPrivateKey << std::endl;
//    
//    
//    
//
//    return 0;
//}

File: server/CryptoManager.py
import base64
import os
from Crypto.PublicKey import RSA
from Crypto.Cipher import AES, PKCS1_OAEP
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad
from cksum import memcrc as get_CRC
class CryptoManager:
    def __init__(self, public_key=None):
        self.private_key = None
        self.public_key = None
        self.aes_key = None

        if public_key:
            try:
                #public_key_bytes = base64.b64decode(public_key)
                self.public_key = RSA.import_key(public_key)
            except (ValueError, TypeError) as e:
                print(f"Invalid public key format: {e}")
    
    def get_encrypted_aes_key(self, public_key = None):
        if public_key:
            self.public_key = public_key
        
        return self.encrypt_rsa(self.aes_key)

    def generate_rsa_keys(self):
        key = RSA.generate(2048)  # Increased key size for better security
        self.private_key = key
        self.public_key = key.publickey()

    def encrypt_aes(self, data):
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        padded_data = pad(data, AES.block_size)
        cipher_aes = AES.new(self.aes_key, AES.MODE_CBC)
        ciphertext = cipher_aes.encrypt(padded_data)
        
        return cipher_aes.iv + ciphertext

    def decrypt_aes(self, enc_data):
        iv = enc_data[:AES.block_size]
        ciphertext = enc_data[AES.block_size:]
        
        cipher_aes = AES.new(self.aes_key, AES.MODE_CBC, iv)
        padded_plaintext = cipher_aes.decrypt(ciphertext)
        plaintext = unpad(padded_plaintext, AES.block_size)
        
        return plaintext

    def encrypt_rsa(self, data):
        if isinstance(data, str):
            data = data.encode('utf-8')
        cipher_rsa = PKCS1_OAEP.new(self.public_key)
        encrypted_data = cipher_rsa.encrypt(data)
        return encrypted_data

    def decrypt_rsa(self, encrypted_data):
        cipher_rsa = PKCS1_OAEP.new(self.private_key)
        decrypted_data = cipher_rsa.decrypt(encrypted_data)
        return decrypted_data

    def generate_aes_key(self):
        self.aes_key = get_random_bytes(32)

    def get_CRC(self, data, decrypt = False):
        if decrypt:
            data = self.decrypt_aes(data)
        
        return get_CRC(data)
    
    
        

# Example usage
if __name__ == "__main__":
    crypto_manager = CryptoManager()
    crypto_manager.generate_rsa_keys()
    crypto_manager.generate_aes_key()
    data = 'Hello my name is Yoni'

    enc_data = crypto_manager.encrypt_aes(data)
    print(enc_data)
    dec_data = crypto_manager.decrypt_aes(enc_data)
    print(dec_data)
